<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Quantify motifs in an experiment — quantify_motifs • glymotif</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantify motifs in an experiment — quantify_motifs"><meta name="description" content="This function takes a glyexp::experiment(),
and returns a new glyexp::experiment() with motif quantifications.
The new experiment is different from a normal glyexp::experiment() in three ways:
It doesn't have a &quot;glycan_structure&quot; or &quot;glycan_composition&quot; column,
but a &quot;motif&quot; column instead.
The metadata field &quot;exp_type&quot; is &quot;motifomics&quot;, not &quot;glycoproteomics&quot; or glycomics.


You can understand a &quot;motif experiment&quot; in this way:
For glycoproteomics data,
instead of containing the quantification of each glycan on each glycosite,
it now contains the quantification of each motif on each glycosite in each sample.
For glycomics data,
it contains the motif quantification in each sample.
Due to the unified data structure of glyexp::experiment(),
the returned motif experiment can be passed to downstream glycoverse
functions like glystats::gly_ttest() for further statistical analysis.
Also, you can use as_tibble() to convert it to a &quot;tidy&quot; tibble for custom analysis."><meta property="og:description" content="This function takes a glyexp::experiment(),
and returns a new glyexp::experiment() with motif quantifications.
The new experiment is different from a normal glyexp::experiment() in three ways:
It doesn't have a &quot;glycan_structure&quot; or &quot;glycan_composition&quot; column,
but a &quot;motif&quot; column instead.
The metadata field &quot;exp_type&quot; is &quot;motifomics&quot;, not &quot;glycoproteomics&quot; or glycomics.


You can understand a &quot;motif experiment&quot; in this way:
For glycoproteomics data,
instead of containing the quantification of each glycan on each glycosite,
it now contains the quantification of each motif on each glycosite in each sample.
For glycomics data,
it contains the motif quantification in each sample.
Due to the unified data structure of glyexp::experiment(),
the returned motif experiment can be passed to downstream glycoverse
functions like glystats::gly_ttest() for further statistical analysis.
Also, you can use as_tibble() to convert it to a &quot;tidy&quot; tibble for custom analysis."><meta property="og:image" content="https://glycoverse.github.io/glymotif/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glymotif</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/glymotif.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/motif-matching.html">The Science Behind Motif Matching</a></li>
    <li><a class="dropdown-item" href="../articles/n-glycans.html">N-Glycans</a></li>
    <li><a class="dropdown-item" href="../articles/with-exp.html">Working with glyexp</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Quantify motifs in an experiment</h1>

      <div class="d-none name"><code>quantify_motifs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes a <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code>,
and returns a new <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> with motif quantifications.</p>
<p>The new experiment is different from a normal <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> in three ways:</p><ol><li><p>It doesn't have a "glycan_structure" or "glycan_composition" column,
but a "motif" column instead.</p></li>
<li><p>The metadata field "exp_type" is "motifomics", not "glycoproteomics" or glycomics.</p></li>
</ol><p>You can understand a "motif experiment" in this way:
For glycoproteomics data,
instead of containing the quantification of each glycan on each glycosite,
it now contains the quantification of each motif on each glycosite in each sample.
For glycomics data,
it contains the motif quantification in each sample.</p>
<p>Due to the unified data structure of <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code>,
the returned motif experiment can be passed to downstream <code>glycoverse</code>
functions like <code>glystats::gly_ttest()</code> for further statistical analysis.
Also, you can use <code>as_tibble()</code> to convert it to a "tidy" tibble for custom analysis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quantify_motifs</span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span>, alignments <span class="op">=</span> <span class="cn">NULL</span>, ignore_linkages <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-exp">exp<a class="anchor" aria-label="anchor" href="#arg-exp"></a></dt>
<dd><p>A <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object containing glycoproteomics data.</p></dd>


<dt id="arg-motifs">motifs<a class="anchor" aria-label="anchor" href="#arg-motifs"></a></dt>
<dd><p>A character vector of motif names, IUPAC-condensed structure strings,
or a 'glyrepr_structure' object.</p></dd>


<dt id="arg-alignments">alignments<a class="anchor" aria-label="anchor" href="#arg-alignments"></a></dt>
<dd><p>A character vector specifying alignment types for each motif.
Can be a single value (applied to all motifs) or a vector of the same length as motifs.</p></dd>


<dt id="arg-ignore-linkages">ignore_linkages<a class="anchor" aria-label="anchor" href="#arg-ignore-linkages"></a></dt>
<dd><p>A logical value. If <code>TRUE</code>, linkages will be ignored in the comparison.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object containing motif quantifications.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Here is a comprehensive comparison of the input experiment and the returned experiment:</p>
<p><strong>Sample information:</strong>
Sample information doesn't change.</p>
<p><strong>Variable information:</strong>
Firstly, we define the "base variables" as the unique combination of all columns
in the variable information tibble except for "variable", "glycan_composition", and "glycan_structure".
The new var_info would be a Cartesian product of all "base variables" and all motifs.
For glycomics data, no "base variable" is defined,
so the resulted var_info only contains the "motif" column.</p>
<p><strong>Expression matrix:</strong>
Rows are variables, columns are samples, same as before.
Filling the cells are the quantifications for motifs on different glycosites in each sample,
instead of quantifications of glycoforms or glycopeptides.
Or just motif quantification in each sample, for glycomics data.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="count_motif.html">count_motifs()</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

