<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Working with glyexp • glymotif</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working with glyexp">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glymotif</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/glymotif.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/motif-matching.html">The Science Behind Motif Matching</a></li>
    <li><a class="dropdown-item" href="../articles/with-exp.html">Working with glyexp</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Working with glyexp</h1>
            
      

      <div class="d-none name"><code>with-exp.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="when-motifs-meet-experiments-a-perfect-partnership">When Motifs Meet Experiments: A Perfect Partnership 🤝<a class="anchor" aria-label="anchor" href="#when-motifs-meet-experiments-a-perfect-partnership"></a>
</h2>
<p>The real power of <code>glymotif</code> shines brightest when it
joins forces with other tools in the <code>glycoverse</code> ecosystem.
If you’re already using <code>glyread</code> to import your
glycoproteomics results and <code>glyexp</code> to manage your
experimental data, you’re in for a treat! <code>glymotif</code> provides
some incredibly useful functions to perform <code>experiment</code>
manipulation about motifs—think of it as adding a new lens to your
analytical microscope. 🔬</p>
<p><strong>Important note:</strong> This vignette assumes you’re
familiar with the <code>glyexp</code> package. If you haven’t met it
yet, we highly recommend checking out its <a href="https://glycoverse.github.io/glyexp/articles/glyexp.html" class="external-link">introduction</a>
first. Trust us—it’s worth the detour! 🚀</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glymotif/">glymotif</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyexp/" class="external-link">glyexp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:glyexp':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select_var</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="meet-our-star-player-real-glycoproteomics-data">Meet Our Star Player: Real Glycoproteomics Data 🌟<a class="anchor" aria-label="anchor" href="#meet-our-star-player-real-glycoproteomics-data"></a>
</h2>
<p>Time to roll up our sleeves and dive into the good stuff! 💪 Let’s
work with a real-world dataset that will showcase what
<code>glymotif</code> can do when it teams up with actual experimental
data. We will use <code>real_experiment</code> from the
<code>glyexp</code> package, an serum N-glycoproteomics study with 12
samples. Firstly, let’s use <code><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">glyclean::auto_clean()</a></code> to
preprocess the data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyclean/" class="external-link">glyclean</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'glyclean'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     aggregate</span></span>
<span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">auto_clean</a></span><span class="op">(</span><span class="va">real_experiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Normalizing data (Median)</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Normalizing data (Median) <span style="color: #B2B2B2;">[162ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Removing variables with &gt;50% missing values</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Removing variables with &gt;50% missing values <span style="color: #B2B2B2;">[46ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Imputing missing values</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample size &lt;= 30, using sample minimum imputation</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Imputing missing values<span style="color: #00BB00;">✔</span> Imputing missing values <span style="color: #B2B2B2;">[25ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Aggregating data</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Aggregating data <span style="color: #B2B2B2;">[895ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Normalizing data again</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Normalizing data again <span style="color: #B2B2B2;">[16ms]</span></span></span>
<span><span class="va">exp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">3880</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">glycan_composition</span> <span style="color: #0000BB;">&lt;comp&gt;</span>, <span style="color: #00BB00;">glycan_structure</span> <span style="color: #0000BB;">&lt;struct&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span></span></span></code></pre></div>
<p>Now, let’s peek under the hood and see what treasures we’re working
with! 👀</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,880 × 6</span></span></span>
<span><span class="co">#&gt;    variable protein gene     glycan_composition     </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;comp&gt;</span>                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> V1       P08185  SERPINA6 Hex(5)HexNAc(4)NeuAc(2)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> V2       P04196  HRG      Hex(5)HexNAc(4)NeuAc(1)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> V3       P04196  HRG      Hex(5)HexNAc(4)        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> V4       P04196  HRG      Hex(5)HexNAc(4)NeuAc(1)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> V5       P10909  CLU      Hex(6)HexNAc(5)        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> V6       P04196  HRG      Hex(5)HexNAc(4)NeuAc(2)</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> V7       P04196  HRG      Hex(5)HexNAc(4)        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> V8       P04196  HRG      Hex(5)HexNAc(4)dHex(2) </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> V9       P04196  HRG      Hex(4)HexNAc(3)        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> V10      P04196  HRG      Hex(4)HexNAc(4)NeuAc(1)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,870 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: glycan_structure &lt;struct&gt;, protein_site &lt;int&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_sample_info.html" class="external-link">get_sample_info</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 2</span></span></span>
<span><span class="co">#&gt;    sample group</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> C1     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> C2     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> C3     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> H3     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> M1     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> M2     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> M3     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Y1     Y    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> Y2     Y    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> Y3     Y</span></span></code></pre></div>
<p>What we have here is a beautiful N-glycoproteomics dataset featuring
500 PSMs (Peptide Spectrum Matches) across 12 samples — a perfect
playground for motif analysis! 🎮</p>
<p><strong>Pro tip:</strong> 💡 In real-world data analysis, you’ll
definitely want to use <code>glyclean</code> to perform data
preprocessing before diving into any analysis. Think of it as washing
your vegetables before cooking—essential for the best results! 🥕</p>
</div>
<div class="section level2">
<h2 id="adding-motif-annotations-to-an-experiment">Adding Motif Annotations to an Experiment 🏷️<a class="anchor" aria-label="anchor" href="#adding-motif-annotations-to-an-experiment"></a>
</h2>
<p>Here’s where things get interesting! 🤔 We know that the variable
information tibble contains all the juicy details about each glycoform -
the proteins, sites, and glycan structures. But what if we want to
sprinkle some motif magic into the mix? What if we want to add more
columns that tell us about the motifs hiding in our glycans?</p>
<p><strong>The simple approach:</strong> One motif at a time (very
intuitive!) 🎯</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/mutate_obs.html" class="external-link">mutate_var</a></span><span class="op">(</span>n_hex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"Hex(a1-"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variable</span>, <span class="va">protein</span>, <span class="va">glycan_structure</span>, <span class="va">n_hex</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,880 × 4</span></span></span>
<span><span class="co">#&gt;    variable protein</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> V1       P08185 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> V2       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> V3       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> V4       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> V5       P10909 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> V6       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> V7       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> V8       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> V9       P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> V10      P04196 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,870 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: glycan_structure &lt;struct&gt;, n_hex &lt;lgl&gt;</span></span></span></code></pre></div>
<p><strong>The tempting approach:</strong> Multiple motifs at once (you
might be tempted to do this…) 🤷‍♀️</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Don't do this</span></span>
<span><span class="va">exp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/mutate_obs.html" class="external-link">mutate_var</a></span><span class="op">(</span></span>
<span>    n_hex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"Hex(a1-"</span><span class="op">)</span>,</span>
<span>    n_hexna <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"HexNAc(a1-"</span><span class="op">)</span>,</span>
<span>    n_dhex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"dHex(a1-"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Hold on there, speed racer! 🛑 While this approach works, it’s not
very efficient, and your computer won’t thank you for it. Here’s why:
those three separate calls to <code><a href="../reference/have_motif.html">have_motif()</a></code> all perform
time-consuming validations and conversions on the same set of glycan
structures. It’s like washing the same dishes three times instead of
doing them all at once! 🍽️ Plus, there’s a lot of repetitive typing. You
have to type <code>have_motif</code> and <code>glycan_structure</code>
three times — talk about finger fatigue! 😴</p>
<p><strong>The smart approach:</strong> Use the
<code><a href="../reference/add_motifs_int.html">add_motifs_lgl()</a></code> or <code><a href="../reference/add_motifs_int.html">add_motifs_int()</a></code> functions
instead! ⚡ They might look like simple syntactic sugar, but they’re
actually optimized powerhouses designed specifically for this exact
scenario:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="va">exp</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/add_motifs_int.html">add_motifs_lgl</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>motif1 <span class="op">=</span> <span class="st">"Hex(??-"</span>, motif2 <span class="op">=</span> <span class="st">"HexNAc(??-"</span>, motif3 <span class="op">=</span> <span class="st">"dHex(??-"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Voilà! 🎉 The motif annotations are now seamlessly integrated into
your variable information tibble.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variable</span>, <span class="va">motif1</span>, <span class="va">motif2</span>, <span class="va">motif3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,880 × 4</span></span></span>
<span><span class="co">#&gt;    variable motif1 motif2 motif3</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> V1       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> V2       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> V3       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> V4       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> V5       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> V6       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> V7       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> V8       TRUE   TRUE   TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> V9       TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> V10      TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,870 more rows</span></span></span></code></pre></div>
<p>But wait, what can you actually <em>do</em> with these shiny new
columns? 🤔 The possibilities are endless, but here’s a tantalizing
example to get your creative juices flowing:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can perform pathway enrichment on all glycoproteins containing some motif:</span></span>
<span><span class="va">exp2</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/filter_obs.html" class="external-link">filter_var</a></span><span class="op">(</span><span class="va">motif1</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">gly_enrich_reactome</span><span class="op">(</span><span class="op">)</span>  <span class="co"># from the `glystats` package</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-art-of-motif-quantification-in-experiments">The Art of Motif Quantification in Experiments 📊<a class="anchor" aria-label="anchor" href="#the-art-of-motif-quantification-in-experiments"></a>
</h2>
<p>Now we’re entering truly exciting territory! 🚀 Meet
<code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> — another incredibly useful tool in our
arsenal. This function takes a <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object
and returns a brand new one with motif quantifications. But what does
that actually <em>mean</em>? 🤷‍♂️</p>
<p>Here’s the beautiful concept: when we quantify a motif in a
glycoproteomic context, we’re measuring the abundance of motifs on each
glycosite in each sample. This information elegantly combines two key
pieces: the motif count on each site <em>and</em> the abundance of that
site. It’s like knowing not just how many sports cars are in a parking
lot, but also how often each parking spot is used! 🚗</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"NeuAc(??-?)Hex(??-?)HexNAc(??-"</span>,</span>
<span>  <span class="st">"Hex(??-?)HexNAc(??-"</span>,</span>
<span>  <span class="st">"HexNAc(??-"</span>,</span>
<span>  <span class="st">"NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,</span>
<span>  <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,</span>
<span>  <span class="st">"dHex(??-?)HexNAc(??-"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">motif_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify_motifs.html">quantify_motifs</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span>, alignments <span class="op">=</span> <span class="st">"terminal"</span>, ignore_linkages <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Pro tip: Our motifs don't have linkage information,</span></span>
<span><span class="co"># but we can still set `ignore_linkages` to TRUE manually to turbocharge the calculation! ⚡</span></span></code></pre></div>
<p>Here’s the beautiful part: <code>motif_exp</code> is also a
<code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code>! 🎭 Think of it this way： if we
describe a typical glycoproteomics <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">experiment()</a></code> as “the
quantities of glycans on different glycosites in different samples”,
then <code>motif_exp</code> is “the quantities of motifs on different
glycosites in different samples”. It’s the same elegant structure, just
viewed through a different analytical lens! 🔍</p>
<p>Let’s explore our newly minted experiment and see what treasures it
holds:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, let's peek at the variable information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="va">motif_exp</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,644 × 5</span></span></span>
<span><span class="co">#&gt;    variable protein protein_site motif                                     gene </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> V1       A6NJW9            49 NeuAc(??-?)Hex(??-?)HexNAc(??-            CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> V2       A6NJW9            49 Hex(??-?)HexNAc(??-                       CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> V3       A6NJW9            49 HexNAc(??-                                CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> V4       A6NJW9            49 NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(?… CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> V5       A6NJW9            49 Hex(??-?)[dHex(??-?)]HexNAc(??-           CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> V6       A6NJW9            49 dHex(??-?)HexNAc(??-                      CD8B2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> V7       O14786           150 NeuAc(??-?)Hex(??-?)HexNAc(??-            NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> V8       O14786           150 Hex(??-?)HexNAc(??-                       NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> V9       O14786           150 HexNAc(??-                                NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> V10      O14786           150 NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(?… NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> V11      O14786           150 Hex(??-?)[dHex(??-?)]HexNAc(??-           NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> V12      O14786           150 dHex(??-?)HexNAc(??-                      NRP1 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,632 more rows</span></span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># And now, the expression matrix (where the magic numbers live!)</span></span>
<span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_expr_mat.html" class="external-link">get_expr_mat</a></span><span class="op">(</span><span class="va">motif_exp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;           C1       C2       C3       H1       H2       H3       M1         M2</span></span>
<span><span class="co">#&gt; V1   4518941  3597488  9291116  1652455 17718.79 18276.18 10131349  4903136.0</span></span>
<span><span class="co">#&gt; V2         0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V3         0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V4   4518941  3597488  9291116  1652455 17718.79 18276.18 10131349  4903136.0</span></span>
<span><span class="co">#&gt; V5         0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V6  13556823 10792465 27873349  4957366 53156.37 54828.54 30394046 14709407.9</span></span>
<span><span class="co">#&gt; V7   8358832 12224955 13576169 13126524 17718.79 18276.18 16723916   904568.5</span></span>
<span><span class="co">#&gt; V8   8358832 12224955 13576169 13126524 17718.79 18276.18 16723916   904568.5</span></span>
<span><span class="co">#&gt; V9         0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V10        0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V11        0        0        0        0     0.00     0.00        0        0.0</span></span>
<span><span class="co">#&gt; V12  8358832 12224955 13576169 13126524 17718.79 18276.18 16723916   904568.5</span></span>
<span><span class="co">#&gt;          M3         Y1       Y2       Y3</span></span>
<span><span class="co">#&gt; V1  2876921   697401.2  3718507  3636137</span></span>
<span><span class="co">#&gt; V2        0        0.0        0        0</span></span>
<span><span class="co">#&gt; V3        0        0.0        0        0</span></span>
<span><span class="co">#&gt; V4  2876921   697401.2  3718507  3636137</span></span>
<span><span class="co">#&gt; V5        0        0.0        0        0</span></span>
<span><span class="co">#&gt; V6  8630764  2092203.5 11155520 10908410</span></span>
<span><span class="co">#&gt; V7  3096274 10234681.7  9805953 12449337</span></span>
<span><span class="co">#&gt; V8  3096274 10234681.7  9805953 12449337</span></span>
<span><span class="co">#&gt; V9        0        0.0        0        0</span></span>
<span><span class="co">#&gt; V10       0        0.0        0        0</span></span>
<span><span class="co">#&gt; V11       0        0.0        0        0</span></span>
<span><span class="co">#&gt; V12 3096274 10234681.7  9805953 12449337</span></span></code></pre></div>
<p>Pretty familiar, right? 😊 It’s just like the experiment you know and
love, except now we’re quantifying motifs rather than glycans.</p>
<p>Here’s the best part: this means <em>all</em> your favorite functions
in <code>glystats</code> can be used on this experiment! 🎉 It’s like
having a universal remote that works with your new TV： everything just
clicks! 📺</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">glystats</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gly_anova</span><span class="op">(</span><span class="va">motif_exp</span><span class="op">)</span></span>
<span><span class="fu">gly_pca</span><span class="op">(</span><span class="va">motif_exp</span><span class="op">)</span></span>
<span><span class="co"># And others...</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
