<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Working with glyexp â€¢ glymotif</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working with glyexp">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glymotif</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.12.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/glymotif.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/motif-matching.html">The Science Behind Motif Matching</a></li>
    <li><a class="dropdown-item" href="../articles/with-exp.html">Working with glyexp</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/glycoverse/glymotif/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Working with glyexp</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/glycoverse/glymotif/blob/main/vignettes/with-exp.Rmd" class="external-link"><code>vignettes/with-exp.Rmd</code></a></small>
      <div class="d-none name"><code>with-exp.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="when-motifs-meet-experiments-a-perfect-partnership">When Motifs Meet Experiments: A Perfect Partnership ğŸ¤<a class="anchor" aria-label="anchor" href="#when-motifs-meet-experiments-a-perfect-partnership"></a>
</h2>
<p>The real power of <code>glymotif</code> shines brightest when it
joins forces with other tools in the <code>glycoverse</code> ecosystem.
If youâ€™re already using <code>glyread</code> to import your
glycoproteomics results and <code>glyexp</code> to manage your
experimental data, youâ€™re in for a treat! <code>glymotif</code> provides
some incredibly useful functions to perform <code>experiment</code>
manipulation about motifsâ€”think of it as adding a new lens to your
analytical microscope. ğŸ”¬</p>
<p><strong>Important note:</strong> This vignette assumes youâ€™re
familiar with the <code>glyexp</code> package. If you havenâ€™t met it
yet, we highly recommend checking out its <a href="https://glycoverse.github.io/glyexp/articles/glyexp.html" class="external-link">introduction</a>
first. Trust usâ€”itâ€™s worth the detour! ğŸš€</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glymotif/">glymotif</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyexp/" class="external-link">glyexp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:glyexp':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select_var</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="meet-our-star-player-real-glycoproteomics-data">Meet Our Star Player: Real Glycoproteomics Data ğŸŒŸ<a class="anchor" aria-label="anchor" href="#meet-our-star-player-real-glycoproteomics-data"></a>
</h2>
<p>Time to roll up our sleeves and dive into the good stuff! ğŸ’ª Letâ€™s
work with a real-world dataset that will showcase what
<code>glymotif</code> can do when it teams up with actual experimental
data. We will use <code>real_experiment</code> from the
<code>glyexp</code> package, an serum N-glycoproteomics study with 12
samples. Firstly, letâ€™s use <code><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">glyclean::auto_clean()</a></code> to
preprocess the data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyclean/" class="external-link">glyclean</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'glyclean'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     aggregate</span></span>
<span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">auto_clean</a></span><span class="op">(</span><span class="va">real_experiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Normalizing data</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> No QC samples found. Using default normalization method based on experiment type.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Experiment type is <span style="color: #0000BB;">"glycoproteomics"</span>. Using `normalize_median()`.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Normalization completed.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Removing variables with too many missing values</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> No QC samples found. Using all samples.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Applying preset <span style="color: #0000BB;">"discovery"</span>...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Total removed: <span style="color: #0000BB;">24</span> (<span style="color: #0000BB;">0.56</span>%) variables.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Variable removal completed.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Imputing missing values</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> No QC samples found. Using default imputation method based on sample size.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Sample size &lt;= 30, using `impute_sample_min()`.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Imputation completed.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Aggregating data</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Aggregating to <span style="color: #0000BB;">"gfs"</span> level</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Aggregation completed.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Normalizing data again</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> No QC samples found. Using default normalization method based on experiment type.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Experiment type is <span style="color: #0000BB;">"glycoproteomics"</span>. Using `normalize_median()`.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Normalization completed.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ <span style="font-weight: bold;">Correcting batch effects</span> â”€â”€</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Batch column  not found in sample_info. Skipping batch correction.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">âœ”</span> Batch correction completed.</span></span>
<span><span class="va">exp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â”€â”€</span> <span style="font-weight: bold;">Glycoproteomics Experiment</span> <span style="color: #00BBBB;">â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">3979</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">â„¹</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">glycan_composition</span> <span style="color: #0000BB;">&lt;comp&gt;</span>, <span style="color: #00BB00;">glycan_structure</span> <span style="color: #0000BB;">&lt;struct&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span></code></pre></div>
<p>Now, letâ€™s peek under the hood and see what treasures weâ€™re working
with! ğŸ‘€</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,979 Ã— 6</span></span></span>
<span><span class="co">#&gt;    variable       protein glycan_composition glycan_structure protein_site gene </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;comp&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;struct&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> P08185-N176-Hâ€¦ P08185  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">Ne</span>â€¦ <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(â€¦          176 SERPâ€¦</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">Ne</span>â€¦ <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)    <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">Ne</span>â€¦ <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> P10909-N291-Hâ€¦ P10909  <span style="color: #000000;">Hex</span>(6)<span style="color: #000000;">HexNAc</span>(5)    <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>â€¦          291 CLU  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">Ne</span>â€¦ <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> P04196-N345-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)    <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>â€¦          345 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(5)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">dH</span>â€¦ <span style="color: #000000;">dHex</span>(??-?)<span style="color: #000000;">Hex</span>(?â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(4)<span style="color: #000000;">HexNAc</span>(3)    <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> P04196-N344-Hâ€¦ P04196  <span style="color: #000000;">Hex</span>(4)<span style="color: #000000;">HexNAc</span>(4)<span style="color: #000000;">Ne</span>â€¦ <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(â€¦          344 HRG  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 3,969 more rows</span></span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_sample_info.html" class="external-link">get_sample_info</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 Ã— 2</span></span></span>
<span><span class="co">#&gt;    sample group</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> C1     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> C2     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> C3     C    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> H1     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> H2     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> H3     H    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> M1     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> M2     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> M3     M    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Y1     Y    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> Y2     Y    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> Y3     Y</span></span></code></pre></div>
<p>What we have here is a beautiful N-glycoproteomics dataset featuring
500 PSMs (Peptide Spectrum Matches) across 12 samples â€” a perfect
playground for motif analysis! ğŸ®</p>
<p><strong>Pro tip:</strong> ğŸ’¡ In real-world data analysis, youâ€™ll
definitely want to use <code>glyclean</code> to perform data
preprocessing before diving into any analysis. Think of it as washing
your vegetables before cookingâ€”essential for the best results! ğŸ¥•</p>
</div>
<div class="section level2">
<h2 id="adding-motif-annotations-to-an-experiment">Adding Motif Annotations to an Experiment ğŸ·ï¸<a class="anchor" aria-label="anchor" href="#adding-motif-annotations-to-an-experiment"></a>
</h2>
<p>Hereâ€™s where things get interesting! ğŸ¤” We know that the variable
information tibble contains all the juicy details about each glycoform -
the proteins, sites, and glycan structures. But what if we want to
sprinkle some motif magic into the mix? What if we want to add more
columns that tell us about the motifs hiding in our glycans?</p>
<p><strong>The simple approach:</strong> One motif at a time (very
intuitive!) ğŸ¯</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/mutate_obs.html" class="external-link">mutate_var</a></span><span class="op">(</span>n_hex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"Hex(a1-"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variable</span>, <span class="va">protein</span>, <span class="va">glycan_structure</span>, <span class="va">n_hex</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,979 Ã— 4</span></span></span>
<span><span class="co">#&gt;    variable                              protein glycan_structure          n_hex</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;struct&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> P08185-N176-Hex(5)HexNAc(4)NeuAc(2)   P08185  <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexN</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(1)-1 P04196  <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexN</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> P04196-N344-Hex(5)HexNAc(4)           P04196  <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(1)-2 P04196  <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexN</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> P10909-N291-Hex(6)HexNAc(5)-1         P10909  <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(2)   P04196  <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexN</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> P04196-N345-Hex(5)HexNAc(4)           P04196  <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> P04196-N344-Hex(5)HexNAc(4)dHex(2)    P04196  <span style="color: #000000;">dHex</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNA</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> P04196-N344-Hex(4)HexNAc(3)-1         P04196  <span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> P04196-N344-Hex(4)HexNAc(4)NeuAc(1)   P04196  <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexN</span>â€¦ FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 3,969 more rows</span></span></span></code></pre></div>
<p><strong>The tempting approach:</strong> Multiple motifs at once (you
might be tempted to do thisâ€¦) ğŸ¤·â€â™€ï¸</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Don't do this</span></span>
<span><span class="va">exp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/mutate_obs.html" class="external-link">mutate_var</a></span><span class="op">(</span></span>
<span>    n_hex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"Hex(a1-"</span><span class="op">)</span>,</span>
<span>    n_hexna <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"HexNAc(a1-"</span><span class="op">)</span>,</span>
<span>    n_dhex <span class="op">=</span> <span class="fu"><a href="../reference/have_motif.html">have_motif</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"dHex(a1-"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Hold on there, speed racer! ğŸ›‘ While this approach works, itâ€™s not
very efficient, and your computer wonâ€™t thank you for it. Hereâ€™s why:
those three separate calls to <code><a href="../reference/have_motif.html">have_motif()</a></code> all perform
time-consuming validations and conversions on the same set of glycan
structures. Itâ€™s like washing the same dishes three times instead of
doing them all at once! ğŸ½ï¸ Plus, thereâ€™s a lot of repetitive typing. You
have to type <code>have_motif</code> and <code>glycan_structure</code>
three times â€” talk about finger fatigue! ğŸ˜´</p>
<p><strong>The smart approach:</strong> Use the
<code><a href="../reference/add_motifs_int.html">add_motifs_lgl()</a></code> or <code><a href="../reference/add_motifs_int.html">add_motifs_int()</a></code> functions
instead! âš¡ They might look like simple syntactic sugar, but theyâ€™re
actually optimized powerhouses designed specifically for this exact
scenario:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="va">exp</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/add_motifs_int.html">add_motifs_lgl</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>motif1 <span class="op">=</span> <span class="st">"Hex(??-"</span>, motif2 <span class="op">=</span> <span class="st">"HexNAc(??-"</span>, motif3 <span class="op">=</span> <span class="st">"dHex(??-"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>VoilÃ ! ğŸ‰ The motif annotations are now seamlessly integrated into
your variable information tibble.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variable</span>, <span class="va">motif1</span>, <span class="va">motif2</span>, <span class="va">motif3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,979 Ã— 4</span></span></span>
<span><span class="co">#&gt;    variable                              motif1 motif2 motif3</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                 <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> P08185-N176-Hex(5)HexNAc(4)NeuAc(2)   TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(1)-1 TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> P04196-N344-Hex(5)HexNAc(4)           TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(1)-2 TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> P10909-N291-Hex(6)HexNAc(5)-1         TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> P04196-N344-Hex(5)HexNAc(4)NeuAc(2)   TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> P04196-N345-Hex(5)HexNAc(4)           TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> P04196-N344-Hex(5)HexNAc(4)dHex(2)    TRUE   TRUE   TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> P04196-N344-Hex(4)HexNAc(3)-1         TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> P04196-N344-Hex(4)HexNAc(4)NeuAc(1)   TRUE   TRUE   FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 3,969 more rows</span></span></span></code></pre></div>
<p>But wait, what can you actually <em>do</em> with these shiny new
columns? ğŸ¤” The possibilities are endless, but hereâ€™s a tantalizing
example to get your creative juices flowing:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can perform pathway enrichment on all glycoproteins containing some motif:</span></span>
<span><span class="va">exp2</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/filter_obs.html" class="external-link">filter_var</a></span><span class="op">(</span><span class="va">motif1</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">gly_enrich_reactome</span><span class="op">(</span><span class="op">)</span>  <span class="co"># from the `glystats` package</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-art-of-motif-quantification-in-experiments">The Art of Motif Quantification in Experiments ğŸ“Š<a class="anchor" aria-label="anchor" href="#the-art-of-motif-quantification-in-experiments"></a>
</h2>
<p>Want to quantify the motifs in your experiment? Try the
<code>glydet</code> package! It provides the
<code>quantify_motifs()</code> function to perform relative and absolute
motif quantification.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
